### PR TITLE
Storage extension for deep copy of project files

### PR DESCRIPTION
**Pull request by mguidon**
## What do these changes do?
Extension of storage api:

- Implements deep clone of a project on storage level (i.e. copying of all files related to a project into internal simcore.s3 storage and, if any, from datcore to simcore.s3)
- Implements deletion of files in internal simcore.s3 storage

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [ ] **Runs in the swarm**
- [ ] If you design a new module, add your user to .github/CODEOWNERS

### CODE REVIEW

File: services/storage/src/simcore_service_storage/dsm.py [hunk 1]
```diff
@@ -426,3 +426,124 @@ def location_from_id(cls, location_id : str):
...
+                            copy_source = {'Bucket' : self.simcore_bucket_name, 'Key': source_object_name}
+                            response = await client.copy_object(CopySource=copy_source, Bucket=self.simcore_bucket_name, Key=dest_object_name)
+
+
+            # Step 2: List all references in outputs that point to datcore and copy over
+            for node_id, node in destination_project['workbench'].items():
+                outputs = node.get("outputs")
+                if outputs is not None:
+                    for _output_key, output in outputs.items():
+                        if "store" in output and output["store"]==1:
```
Comment (pcrespov) [inline services/storage/src/simcore_service_storage/dsm.py]: Do not user 1 or 0, Please use instead [IntEnum](https://docs.python.org/3/library/enum.html#intenum), then e.g.
``python
output["store"]==int(Store.DATCORE)
``

or even better, just use ``DATCORE_ID`` in [settings.py](services/storage/src/simcore_service_storage/settings.py).  Or adapt thos to an IntEnum. Consider that ``Store.DATCORE.name == DATCORE``

File: services/storage/src/simcore_service_storage/dsm.py [hunk 2]
```diff
@@ -426,3 +426,124 @@ def location_from_id(cls, location_id : str):
...
+                            output["store"] = 0
+                            output['path'] = dest
+
+        # step 3: list files first to create fmds
+        session = aiobotocore.get_session(loop=_loop)
+        fmds = []
+        async with session.create_client('s3', endpoint_url=self.s3_client.endpoint_url, aws_access_key_id=self.s3_client.access_key,
+            aws_secret_access_key=self.s3_client.secret_key) as client:
+            response = await client.list_objects_v2(Bucket=self.simcore_bucket_name, Prefix=dest_folder+"/")
+            if 'Contents' in response:
```
Comment (pcrespov) [inline services/storage/src/simcore_service_storage/dsm.py]: If no contents, shouldn;t you raise exception?

File: services/storage/src/simcore_service_storage/dsm.py [hunk 3]
```diff
@@ -426,3 +426,124 @@ def location_from_id(cls, location_id : str):
...
+        async with session.create_client('s3', endpoint_url=self.s3_client.endpoint_url, aws_access_key_id=self.s3_client.access_key,
+            aws_secret_access_key=self.s3_client.secret_key) as client:
+            response = await client.list_objects_v2(Bucket=self.simcore_bucket_name, Prefix=source_folder)
+
+            if "Contents" in response:
+                for f in response['Contents']:
+                    source_object_name = f['Key']
+                    source_object_parts = Path(source_object_name).parts
+
+                    if len(source_object_parts) == 3:
```
Comment (pcrespov) [inline services/storage/src/simcore_service_storage/dsm.py]: Error if not content or if not len(source_object_parts) == 3??
At least log int?

File: services/storage/src/simcore_service_storage/handlers.py [hunk 4]
```diff
@@ -246,6 +246,47 @@
...
+    await dsm.deep_copy_project_simcore_s3(user_id, source_project, destination_project, nodes_map)
+
+    raise web.HTTPCreated(text=json.dumps(destination_project),
+                                content_type='application/json')
+
+async def delete_folders_of_project(request: web.Request):
+    folder_id = request.match_info['folder_id']
+    user_id = request.query.get("user_id")
+
+    # TODO: implement
```
Comment (pcrespov) [inline services/storage/src/simcore_service_storage/handlers.py]: remoev comments

File: services/storage/tests/test_dsm.py [hunk 5]
```diff
@@ -364,3 +366,116 @@ def test_fmd_build():
...
+        destination_project["workbench"][key] = node
+        node_mapping[node_id] = key
+
+    status = await dsm.deep_copy_project_simcore_s3(user_id, source_project, destination_project, node_mapping)
+    new_path = destination_project["workbench"]["deep-copy-uuid-48eb-a9d2-aaad6b72400a"]["outputs"]["outFile"]["path"]
+    assert new_path != path_in_datcore
+    files = await dsm.list_files(user_id=user_id, location=SIMCORE_S3_STR)
+    assert len(files) == 3
+    found = False
+    for f in files:
```
Comment (pcrespov) [inline services/storage/tests/test_dsm.py]: Minor: 
```python
assert any(f.file_uuid == new_path for f in files)
```

File: services/storage/tests/test_dsm.py [hunk 6]
```diff
@@ -364,3 +366,116 @@ def test_fmd_build():
...
+
+    assert found
+
+    response = await dsm.delete_project_simcore_s3(user_id, destination_project["uuid"])
+    print(response)
+
+    files = await dsm.list_files(user_id=user_id, location=SIMCORE_S3_STR)
+    assert len(files) == 0
+
+async def test_mock(mocker, dsm_fixture):
```
Comment (pcrespov) [inline services/storage/tests/test_dsm.py]: I think you can remove this test, right? it is a dev test so I would mark as 

```python
 @pytest.mark.skip(reason-"dev-test")  
```

File: services/storage/src/simcore_service_storage/dsm.py [hunk 7]
```diff
@@ -426,3 +426,124 @@ def location_from_id(cls, location_id : str):
...
+                        old_node_id = source_object_parts[1]
+                        new_node_id = node_mapping.get(old_node_id)
+                        if new_node_id is not None:
+                            old_filename = source_object_parts[2]
+                            dest_object_name = str(Path(dest_folder) / Path(new_node_id) / Path(old_filename))
```
[the remark this answers is not in the document]
Comment (mguidon) [reply services/storage/src/simcore_service_storage/dsm.py]: ok

File: services/storage/tests/test_dsm.py (hunk 5 above)
Comment (mguidon) [reply services/storage/tests/test_dsm.py]: nice

File: services/storage/tests/test_dsm.py (hunk 6 above)
Comment (mguidon) [reply services/storage/tests/test_dsm.py]: true that

File: services/storage/src/simcore_service_storage/handlers.py (hunk 4 above)
Comment (mguidon) [reply services/storage/src/simcore_service_storage/handlers.py]: ok

File: services/storage/src/simcore_service_storage/dsm.py (hunk 1 above)
Comment (mguidon) [reply services/storage/src/simcore_service_storage/dsm.py]: oops

File: services/storage/src/simcore_service_storage/dsm.py (hunk 2 above)
Comment (mguidon) [reply services/storage/src/simcore_service_storage/dsm.py]: I think if there are no objects, there is no content

**Pull request merged**